spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/core/Mqttv5ClientManager.java

```diff
@@ -0,0 +1,147 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.mqtt.core;
+
+import org.eclipse.paho.mqttv5.client.IMqttAsyncClient;
+import org.eclipse.paho.mqttv5.client.IMqttToken;
+import org.eclipse.paho.mqttv5.client.MqttAsyncClient;
+import org.eclipse.paho.mqttv5.client.MqttCallback;
+import org.eclipse.paho.mqttv5.client.MqttConnectionOptions;
+import org.eclipse.paho.mqttv5.client.MqttDisconnectResponse;
+import org.eclipse.paho.mqttv5.common.MqttException;
+import org.eclipse.paho.mqttv5.common.MqttMessage;
+import org.eclipse.paho.mqttv5.common.packet.MqttProperties;
+
+import org.springframework.util.Assert;
+
+public class Mqttv5ClientManager extends AbstractMqttClientManager<IMqttAsyncClient> implements MqttCallback {
+
+	private final MqttConnectionOptions connectionOptions;
+
+	private volatile IMqttAsyncClient client;
+
+	public Mqttv5ClientManager(String url, String clientId) {
+		super(clientId);
+		Assert.notNull(url, "'url' is required");
+		setUrl(url);
+		this.connectionOptions = new MqttConnectionOptions();
+		this.connectionOptions.setServerURIs(new String[]{ url });
+		this.connectionOptions.setAutomaticReconnect(true);
+	}
+
+	public Mqttv5ClientManager(MqttConnectionOptions connectionOptions, String clientId) {
+		super(clientId);
+		Assert.notNull(connectionOptions, "'connectionOptions' is required");
+		this.connectionOptions = connectionOptions;
+		if (!this.connectionOptions.isAutomaticReconnect()) {
+			logger.warn("It is recommended to set 'automaticReconnect' MQTT connection option. " +
+					"Otherwise connection check and reconnect should be done manually.");
+		}
+		Assert.notEmpty(connectionOptions.getServerURIs(), "'serverURIs' must be provided in the 'MqttConnectionOptions'");
+		setUrl(connectionOptions.getServerURIs()[0]);
+	}
+
+	@Override
+	public IMqttAsyncClient getClient() {
+		return this.client;
+	}
+
+	@Override
+	public synchronized void start() {
+		if (this.client == null) {
+			try {
+				this.client = new MqttAsyncClient(getUrl(), getClientId());
+				this.client.setManualAcks(isManualAcks());
+				this.client.setCallback(this);
+			}
+			catch (MqttException e) {
+				throw new IllegalStateException("could not start client manager", e);
+			}
+		}
+		try {
+			this.client.connect(this.connectionOptions)
+					.waitForCompletion(this.connectionOptions.getConnectionTimeout());
+		}
+		catch (MqttException e) {
+			logger.error("could not start client manager, client_id=" + this.client.getClientId(), e);
```

Comment:
I think the `this.mqttClient.reconnect()` has to be done over here.
See similar logic in the `Mqttv5PahoMessageDrivenChannelAdapter.doStart()`

Reply:
Yeah, actually I have one "flaky" test `ClientManagerBackToBackTests.testSameV5ClientIdWorksForPubAndSub` which started failing today - because of exactly this issue! Client failed to connect first time and could not subscribe to the topic.
Thanks, I made it similar to adapter code.